test/framework/src/main/java/org/elasticsearch/test/ClusterServiceUtils.java

```diff
@@ -284,11 +294,7 @@ public String toString() {
         };
         clusterService.addListener(clusterStateListener);
         listener.addListener(ActionListener.running(() -> clusterService.removeListener(clusterStateListener)));
-        if (predicate.test(clusterService.state())) {
-            listener.onResponse(null);
-        } else {
-            listener.addTimeout(ESTestCase.SAFE_AWAIT_TIMEOUT, clusterService.threadPool(), EsExecutors.DIRECT_EXECUTOR_SERVICE);
-        }
+        listener.addTimeout(ESTestCase.SAFE_AWAIT_TIMEOUT, clusterService.threadPool(), EsExecutors.DIRECT_EXECUTOR_SERVICE);
```

Comment:
Checking the current cluster state is important I think, or else we will always have to wait for the next state update (which might not ever happen)

Reply:
I still do that, I just moved it up a few lines to the start of the method. That avoids adding the cluster state listener unnecessarily.

Reply:
Ah but you have to check the state _after_ adding the listener, or else you might miss an update that happens in between checking the state and adding the listener.